### ISSUE TITLE
Using the handle widget to reorder records fails when the sequence field includes large integers

### ISSUE THREAD

**Issue by danielkauffman** [odoo/odoo#39608]
**Impacted versions:**

Odoo 13.

**Steps to reproduce:**

Create an integer field and assign a large integer value to that field. To make it easy to reproduce the issue, assign the max value of PostgreSQL SMALLINT which is a signed 32-bit integer.

> sequence = odoo.fields.Integer(default=int((1 << 31) - 1))

Create a tree view that uses the handle widget:

> \<field name="sequence" widget="handle" /\>

Open the tree view and use the handle to reorder records.

**Current behavior:**

The records are created correctly, but when the handle widget tries to reorder the records, the sequence goes out of range, the transaction is rolled back, and the records are not reordered.

**Expected behavior:**

The handle widget should be able reorder any number of records having arbitrary default values. As long as the sequence value is a valid integer, the handle widget should not fail.

**Links:**

https://github.com/odoo/odoo/blob/13.0/addons/web/controllers/main.py#L1328
https://github.com/odoo/odoo/blob/13.0/addons/web/static/src/js/views/basic/basic_controller.js#L729
https://github.com/odoo/odoo/blob/13.0/addons/web/static/src/js/views/basic/basic_model.js#L940

**Possible fix:**

When the handle widget triggers resequencing, instead of starting at zero, the sequence seems to be starting at an arbitrarily high value, which arbitrarily limits the number of records that can be resequenced. If instead the resequencing always started at zero, this issue would probably be resolved.

**Workaround:**

~~If you want new records to always be added to the end of the recordset, but you don't know how large the recordset might be, set the default sequence position for new records to be an arbitrary value that is larger than the max expected size of the recordset AND smaller than the max value of PostgreSQL SMALLINT minus the expected size of the recordset. Thus the largest default value is int((1<<30)-1).~~

~~This workaround resolves the issue for any practical size of recordset having a handle widget.~~

EDIT: See code below for workaround to add new records in form view with the new records sorted at the end of the list.

**Comment by pedrobaeza** [comment]
This is a very specific thing that it's not going to hit regular code, and you have a workaround for not hitting it. I think the reason is that it's reordering from the previous number and leaving an step higher than 1 for allowing insertions in between.

Any way, I let @seb-odoo to talk about this and see if this is feasible to be changed or should be a known limitation.

**Comment by danielkauffman** [comment]
This issue could also be described as: using the handle widget to reorder records starts from the existing values. As such, there isn't any **constant** that can be used as a default value to allow new records to be sorted at the end of the list.

However, if using a function to calculate the default value, the current max value can be calculated by looping over the recordset to calculate the max value + 1, or using the cursor to get the max value directly from the database, and that max value can be used as the new default, which successfully sorts new records at the end of the list.

```python
def _default_sequence(self):
    """Sort new records at the end of the list."""
    sequence = 0
    for record in self.search([]):
        sequence = max(sequence, record.sequence + 1)
        return sequence

sequence = odoo.fields.Integer(default=_default_sequence)
```

**Comment by pedrobaeza** [comment]
But I don't get why you need to do that default sequence... It's handled automatically by the widget. Just put editable="bottom" or "top" for adding at the beginning or at the end nevertheless the existing values.

**Comment by danielkauffman** [comment]
Great suggestion. Unfortunately, that only works in tree view. I'm wanting to add records in form view, and have the new records sorted at the end of the list.

**Comment by pedrobaeza** [comment]
I think that if you add the widget handle in the tree view, although editing in form view (and having sequence field in form as invisible), you will get the same result.

**Comment by danielkauffman** [comment]
I added the sequence field to the form view, but the new record simply uses the default value for the sequence field. The handle widget doesn't seem to do anything to initialize the field.

I also played with editable="bottom" on the tree view, but that prevents me from accessing the form view.

On a whim, I tried adding editable="bottom" to the form view, but that had no effect.

Thus far, the only way I've found to add new records that are sorted at the end of the list is using the `_default_sequence()` function described above.

**Comment by seb-odoo** [comment]
Indeed the handle widget of the list does not have an impact on form views when they are on different pages (eg. ecommerce categories). The only case where it works is for example on o2m fields where the form view is just a dialog on top of the existing list.

The only way to have the correct sequence for standalone form views would be to have the default value being computed with that kind of `_default_sequence` method. We already do that for some models. (eg. website_sequence on product, sequence on website menu)

A generic default sequence method was investigated there but it didn't go further: https://github.com/odoo/odoo/pull/27634

More investigation on defaults was made to correctly increment the default sequence method for each record when creating in batch: https://github.com/odoo/odoo/pull/36704
It's currently not being worked on, but I still would like to get it some day.

The current solution is to create default sequence method manually for each model that needs it.

None of those actually solve the issue with large integers, but it's just not the scale at which we work I believe. Still for that reason I would advise that the default sequence starts from a low value when possible (eg. for a sales order line, basing the default sequence on the current max sequence related to the same sales order, and not based on the sequence of any other line in the database). I think this complexity is the main reason why the first PR I linked (with the generic solution) didn't go further.

**Comment by pedrobaeza** [comment]
Thanks for the insights. Then we should close as "won't fix", but with the specified ways for workarounding it.

**Comment by mstuttgart** [comment]
> This issue could also be described as: using the handle widget to reorder records starts from the existing values. As such, there isn't any **constant** that can be used as a default value to allow new records to be sorted at the end of the list.
> 
> However, if using a function to calculate the default value, the current max value can be calculated by looping over the recordset to calculate the max value + 1, or using the cursor to get the max value directly from the database, and that max value can be used as the new default, which successfully sorts new records at the end of the list.
> 
> ```python
> def _default_sequence(self):
>     """Sort new records at the end of the list."""
>     sequence = 0
>     for record in self.search([]):
>         sequence = max(sequence, record.sequence + 1)
>         return sequence
> 
> sequence = odoo.fields.Integer(default=_default_sequence)
> ```

I need to make a little fix (odoo 12.0):

```python
def _default_sequence(self):
     """Sort new records at the end of the list."""
     sequence = 0
     for record in self.search([]):
         sequence = max(sequence, record.sequence + 1)
        
     # here: -> fix return indent
     return sequence

sequence = odoo.fields.Integer(default=_default_sequence)
 
```
